keps/0007-dynamic-audit-control.md

```diff
@@ -0,0 +1,126 @@
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-05-18
+last-updated: 2018-05-18
+status: provisional
+see-also:
+  - need to find dynamic admission KEP
+---
+
+# Dynamic Audit Control
+
+## Table of Contents
+
+* [Dynamic Audit Control](#dynamic-audit-control)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+      * [Dynamic Configuration of Policy](#dynamic-configuration-of-policy)
+      * [Dynamic Configuration of Flags](#dynamic-configuration-of-flags)
+      * [User Stories](#user-stories)
+        * [Story 1](#story-1)
+        * [Story 2](#story-2)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Risks and Mitigations](#risks-and-mitigations)
+  * [Graduation Criteria](#graduation-criteria)
+  * [Implementation History](#implementation-history)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+We want to allow the advanced auditing features to be dynamically configured. Following in the same vein as [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) we would like to provide a means of configuring the auditing features post cluster provisioning.
+
+## Motivation
+
+The advanced auditing features are a powerfull tool, yet difficult to configure. The configuration requires deep insight into the deployment mechanism of choice and often takes many iterations to configure properly requiring a restart of the apiserver each time. Moreover, the ability to install addon tools that configure and enhance audting is hindered by the overhead in configuration. Such tools frequently run on the cluster requiring future knowledge of how to reach them when the cluster is live. These tools could enhance the security and conformance of the cluster and its applications.
+
+### Goals
+- Provide an api and set of objects to configure the advanced auditing kube-apiserver webhook flags dynamically
+- Allow the current audit policy object to be configured dynamically
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration of Policy
+Dynamic configuration of the existing policy object will be provided through use of a shared informer.
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: Policy
+omitStages:
+  - "RequestReceived"
+rules:
+  - level: <request level>
+    resources:
+    - group: <group>
+      resources: <resources>
+```
+In kube-apiserver today this is only accepted as a file, this KEP would allow it to either be configured as a file on the master node or applied to the `kube-system` namespace as an object and be dynamically configured. This will require the audit API to be registered with the apiserver. If both file and object exist the dynamic object will take precedence.
+
+### Dynamic Configuration of Flags
```

Comment:
https://goo.gl/opSc2o laid out the general recommendation for kubernetes component configuration
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/dynamic-kubelet-configuration.md is proposal for kubelet

I think kube api server would move towards this in the future. I'd like the change proposed here to be consistent with this direction.

Reply:
+1

I can easily see the policies to be API objects as these are files in the traditional flag based configuration. But the audit plugin flags ought to be in-line with the componentconfig direction.

Reply:
@yliaog thank you for sharing the doc. The main takeaway I'm seeing that applies here is the movement away from static disk configuration. The only motivation for keeping the root policy as an optionally static configuration was to scope down access to admins with rights to the master node. I strongly agree with the sentiments and direction of the doc, but would want the sign off from sig-auth before removing static disk configuration entirely.

Please inform if there are other pieces of this proposal that aren't in line with the doc.

Reply:
along these lines, I'm curious if there is interest in providing a common dynamic webhook interface. Dynamic admission currently has https://github.com/kubernetes/kubernetes/blob/v1.10.0-beta.1/staging/src/k8s.io/api/admissionregistration/v1beta1/types.go#L218 but that lacks the robust auth options that a kubeconfig file offers. Would it make sense for there to be a common interface for outgoing webhooks that can load secrets for its auth infromation?

Reply:
> Would it make sense for there to be a common interface for outgoing webhooks that can load secrets for its auth infromation?

Auth was intentionally omitted from that structure, since it is useful and sometimes necessary to be able to distinguish the apiserver calling the webhook. For example, I would not expect all extension servers to have the ability to identify themselves as the main kube-apiserver when calling webhooks.

Reply:
That makes sense, I wonder if theres still value in providing a common mechanism for how outgoing dynamic webhooks are handled or whether we should duplicate that pattern within this implementation for the time being

Reply:
I'm also seeing the lack of auth here problematic. With the dynamic admission control its not really needed as your simply checking a token, with audit this opens an attack vector of an on cluster agent pushing phony events to the backend.

Reply:
A few followup questions to discuss:
1. what would be ideal config structure? I think this should be considered holistically for API server. we don't want to have a config struct for audit, another config struct for, say, storage, and yet another config for something else. Probably we want to have ApiServerConfiguration as top level struct, and AuditConfiguration as a sub struct
2. There is a difference between MutatingWebhookConfiguration and KubeletConfiguration in that one is a top level k8s API, the other is using ConfigMap. Hence it bears the question of whether AuditConfiguration should be a k8s API object. There is pros/cons of each, it is worthy of further discussion.
3. It seems that even with this dynamic config, there will still be audit related flags configured through command line (is there 30-ish of them?). would it make sense to group them together as a single config struct? and as a single disk file provided to API server.
4. Generally, what is the planned process to rollout config, rollback config? how to report / monitor the status of the rollout / rollback?

Reply:
Here's what I have so far:
1. I think a holistic config makes sense, but with audit attempting to be a dynamic configuration, I'm not sure how those two would play together. We could have a single `APIServer` configuration object, but I fear that may get bloated. We could potentially have a set of configuration objects that are namespaced in a certain way
2. A well defined API object seems to be more clear, what are the advantages of nesting it as a json blob in a configmap?
3. I would hope to get rid of the majority of the flags, or as you said roll them up into a single config object. Ideally that would be a singleton object that could be read from disk or provided via the api.
4. The current solution is to use a shared informer to listen for config changes for the dynamic config, and on rollout log the changes. If the config if invalid, that should be logged, and we could potentially provide a 'last known good config' to roll back to